### DISCUSSION TITLE
Waypoint navigation system

### DISCUSSION THREAD

**Discussion by brats2017** [JSBSim-Team/jsbsim#346] [Support requests]
Hello all,
I want to design a waypoint navigation system for the aircraft x15. I read the manual and I run the c3104.xml file without any problem. However, for designing the x15 waypoint navigation system, I face problems. The extra works that I did to design a waypoint navigation system are as follows:
1) I copied the c3104.xml file from the jsbsim-master\scripts and pasted it with a different name (x15way.xml) in jsbsim-master\scripts.
2) I changed the 8th line of x15way.xml to `<use aircraft="X15" initialize="ellington"/>` and saved it.
3) I copied the ellington.xml file and pasted it in this location: jsbsim-master\aircraft\X15.

When I run the x15way.xml, I have errors that said some properties are not defined. Consequently:

4) I added these properties to the x15ap.xml file and then I run the x15way.xml file. At this step, I do not have any error, but in the command prompt window in front of the distance to the wp at event1, is written 0. Why? 

![waypoint](https://user-images.githubusercontent.com/65332612/99832388-75d02f80-2b2e-11eb-8d85-f00eaf4c6219.png)

The properties that I had to add, are:

```
  <property>ap/attitude_hold</property>
  <property>ap/heading_hold</property>
  <property>ap/yaw_damper</property>
  <property>ap/active-waypoint</property>
  <property>ap/altitude_setpoint</property>
  <property>ap/heading_setpoint</property>
  <property>ap/heading-setpoint-select</property>
  <property>ap/aileron_cmd</property>
  <property>ap/elevator_cmd</property>
  <property>ap/rudder_cmd</property>
  <property>ap/airspeed_setpoint</property>
  <property>ap/airspeed_hold</property>
  <property>ap/throttle-cmd-norm</property>
  <property>guidance/wp-distance</property>
  <property>fcs/mixture-cmd-norm[1]</property>
  <property>fcs/advance-cmd-norm[1]</property>
  <property>propulsion/magneto_cmd</property>
  <property>fcs/throttle-cmd-norm[1]</property>
  <property>fcs/propulsion/starter_cmd</property>
  <property>fcs/propulsion/starter_cmd</property>
  <property>propulsion/starter_cmd</property>
  <property>guidance/target_wp_latitude_rad</property>
  <property>guidance/target_wp_longitude_rad</property>
  <property>ap/altitude_hold</property>
```

I wanted to kindly ask you the steps that I did to design a waypoint navigation system are correct?
I am a beginner with JSBSim and your help will be greatly appreciated.
Thanks in advance

**Comment by bcoconni** [comment]
Hi @brats2017,

The flight controls for waypoint guidance are located in [systems/GNCUtilities.xml](https://github.com/JSBSim-Team/jsbsim/blob/master/systems/GNCUtilities.xml). The c310 loads this autopilot feature with the code below but the X15 does not.
https://github.com/JSBSim-Team/jsbsim/blob/82de86e7dcf6b8029bbbf676dbff99e6bdc30b9d/aircraft/c310/c310.xml#L343-L345

Also please bear in mind that the X15 has very limited range and maneuverability and might not be able to perform the flight plan described in the script `c3104.xml`.

**Edit:** Also the simple fact that there are so many properties missing should be an hint that some files are missing in your model (i.e. they are not included by a command like the one above).

**Comment by brats2017** [comment]
Thank you so much for your quick response @bcoconni. I could design a waypoint navigation system for x15 with the points that you mentioned. Now, I can work on constraints to get good results.

**Comment by brats2017** [comment]
I wanted to kindly ask you how can I consider a constraint for the maximum speed?
Thanks in advance

**Comment by bcoconni** [comment]
> how can I consider a constraint for the maximum speed?

Not sure what you mean. If you want to cap the maximum speed, my suggestion would be to use a PID that would take over the throttle as soon as the aircraft velocity goes beyond a given threshold. Then the PID would be switched off as soon as the velocity drops below that threshold. Just an idea off the top of my head...

**Comment by brats2017** [comment]
Thank you so much for your response @bcoconni.

**Comment by bcoconni** [comment]
@brats2017 according to your discussion in issue #349 I infer that this issue can be closed ?

**Comment by brats2017** [comment]
Yes, sure. Thank you so much @bcoconni.

**Comment by brats2017** [comment]
> > how can I consider a constraint for the maximum speed?
> 
> Not sure what you mean. If you want to cap the maximum speed, my suggestion would be to use a PID that would take over the throttle as soon as the aircraft velocity goes beyond a given threshold. Then the PID would be switched off as soon as the velocity drops below that threshold. Just an idea off the top of my head...

Sorry @bcoconni, the issue is closed. However, my problem has not been solved yet. Can you please give me advice that what are the exact **properties name** (for input and output) that I should consider in the PID controller code to cap the maximum speed?

```
   <pid name="maximum speed">
     <input> **?** </input>
     <kp> value </kp>
     <ki>  value </ki>
     <kd> value </kd>
     <output> **?** </output>
   </pid>
```
Your help will be greatly appreciated.

**Reply by bcoconni** [reply]
> Sorry @bcoconni, the issue is closed.

No problem. I am taking this opportunity to test the new forum feature form GitHub.
> Can you please give me advice that what are the exact properties name (for input and output) that I should consider in the PID controller code to cap the maximum speed?

You would typically have a property that would contain the maximum speed, say `maximum-true-velocity-fps` and compute the error between the current true velocity and the maximum speed.
```xml
<summer name="maximum-speed-error">
  <input> maximum-true-velocity-fps </input>
  <input> -velocities/vt-fps </input>
</summer>
```
Then your PID would drive the engine throttle based on the maximum speed error
```xml
<pid name="maximum speed">
     <input> maximum-speed-error </input>
     <kp> value </kp>
     <ki>  value </ki>
     <kd> value </kd>
     <output> fcs/throttle-cmd-norm[0] </output>
</pid>
```

**Reply by brats2017** [reply]
Thank you so much for your detailed answer @bcoconni. I could control the speed with your help.

**Comment by seanmcleod** [comment]
You do realise that the X-15 is carried aloft under the wing of a B-52, then dropped from altitude, lights its rocket engine which has a limited run time and then becomes a glider.

So you would only be able to use the throttle of the rocket engine for a small portion of the flight to control speed. When it becomes a glider the throttle will be useless, you would need to use alpha (within limits) to control speed. The X-15 also had speed brakes (split tail) which you could use to help in managing speed, but the current X-15 JSBSim model doesn't seem to model the speed brake.

So which part of the flight do you want to limit speed for? If it's just while the rocket is firing then you can use the throttle to control that.

**Comment by brats2017** [comment]
> You do realise that the X-15 is carried aloft under the wing of a B-52, then dropped from altitude, lights its rocket engine which has a limited run time and then becomes a glider.
> 
> So you would only be able to use the throttle of the rocket engine for a small portion of the flight to control speed. When it becomes a glider the throttle will be useless, you would need to use alpha (within limits) to control speed. The X-15 also had speed brakes (split tail) which you could use to help in managing speed, but the current X-15 JSBSim model doesn't seem to model the speed brake.
> 
> So which part of the flight do you want to limit speed for? If it's just while the rocket is firing then you can use the throttle to control that.

Thank you so much for your response. The information that you mentioned is very helpful. About the question that you asked me, I am designing a waypoint navigation system for X-15, and I want to control its speed when it is following the waypoints. Its speed is too high when it heads to the first waypoint and it causes the aircraft to deviate. Do you have any idea about that?
Thanks in advance.

**Comment by seanmcleod** [comment]
You haven't answered the question I asked. 

> So which part of the flight do you want to limit speed for?

You mention the leg to the 1st waypoint, but that doesn't specify whether it's under rocket power or whether it's gliding, and as I mentioned you will need to use a different strategy for the different phases of flight in general.

**Reply by brats2017** [reply]
Thank you so much for your response @seanmcleod. It is under rocket power, and I could control the velocity of aircraft using the throttle as you mentioned. However, my main problem has not been solved yet, because I thought by controlling the speed I could get a result, but it did not help me.

Please let me explain my problem. I wanted to design a waypoint navigation system for X15 in JSBSim, so I imitated the files related to c310 and made some files for X15. The first three events that I considered are: Start engine, Set altitude, and Raise landing gear. The aircraft run these events without error, but it does not consider the conditions that I considered in each event, and it causes the aircraft not to follow other events. I tried a lot to tune the controllers (heading hold, altitude hold) in the autopilot file. However, I could not succeed, and the aircraft could not follow the conditions in these events.

However, with your explanation, I have a question for you. You mentioned that: 

> You do realise that the X-15 is carried aloft under the wing of a B-52, then dropped from altitude, lights its rocket engine which has a limited run time and then becomes a glider.

About designing a waypoint navigation system for X15, do you think that I should consider these events before when I want to ask the aircraft to go to the first waypoint? I mean do you think it is a better way that I start to design a waypoint navigation system when the aircraft is in gliding mode? 

Your help will be greatly appreciated.

**Reply by seanmcleod** [reply]
> should consider these events before when I want to ask the aircraft to go to the first waypoint?

It all depends what your requirements/needs are. So you could design a waypoint controller that works while the aircraft is under rocket power and as a glider or one that only works with rocket power or only as a glider.

But starting out I'd suggest picking one of the options and design the controller for that portion of the flight.

**Reply by bcoconni** [reply]
>  I tried a lot to tune the controllers (heading hold, altitude hold) in the autopilot file. However, I could not succeed, and the aircraft could not follow the conditions in these events.

I may be wrong but it seems you are trying to tune several controllers simultaneously which is not good practice. I would rather tune one controller at a time.

In addition, there already exist altitude and pitch hold controllers for the X15. These are used in the script `x153.xml` which simulates a powered flight from the wing drop to the apogee. If you check the script, you will see that there is an event for a steady altitude flight for 15 seconds (lines 47 to 60):
https://github.com/JSBSim-Team/jsbsim/blob/22d57cd8849fb80c4ab8fcb3f48c15e20b89f9e0/scripts/x153.xml#L1-L118
I would suggest not trying to implement the waypoint controllers or the speed limiter before you succeed having the X15 maintaining a constant altitude powered flight. Only then I would implement the speed limiter and finally the waypoint controller. In that order.

**Reply by brats2017** [reply]
Thank you so much for your detailed answer @bcoconni. I could design a waypoint navigation system for X15 with your help. However, I have a problem like a problem that I have when I run the c3104.xml file. When I run the c3104.xml file, in the second waypoint, the aircraft heads to the target instead of heading to the second waypoint. Actually, about other waypoints, the aircraft heads to the correct waypoints, but for the second waypoint, it does not consider the coordinate of the second waypoint, and it heads to the target. Do you have any suggestions about it?

**Reply by bcoconni** [reply]
I plotted the properties `ap/active-waypoint` and  `guidance/wp-distance`  from the script `c3104.xml`.

![active-waypoint](https://user-images.githubusercontent.com/7502904/102534003-6baa3e00-40a6-11eb-9679-86a959f8b8bf.jpg)
Obviously the autopilot switches to the 2nd waypoint at t = 285 s and then switches to the 3rd waypoint at t=310 s.

![wp-distance](https://user-images.githubusercontent.com/7502904/102533967-5fbe7c00-40a6-11eb-8a98-a86aa98cbce9.jpg)
According to the plot of `guidance/wp-distance`, the aircraft is already quite close to the 2nd waypoint when it is activated so the autopilot follows its direction for a much shorter duration than it does for the 3 other waypoints. This might give the illusion that the 2nd waypoint is skipped ?

Do you find different results ? If so, could you please show the plots you obtain for the properties `ap/active-waypoint` and  `guidance/wp-distance` ?

**Reply by brats2017** [reply]
Thank you so much for your response @bcoconni . I got similar results for ap/active-waypoint and guidance/wp-distance. However, I think these results are not efficient to prove whether the aircraft is heading to the second waypoint or not. The following plot shows the aircraft is heading to the target when the second waypoint is active. Also, when I choose other locations for target and second waypoint, the aircraft heads to the new target (when the second waypoint is active), without considering the second waypoint. Please let me know if you have any idea about that.
![waypoint](https://user-images.githubusercontent.com/65332612/103128940-01286d80-4665-11eb-8863-ad3d592d9302.jpg)

**Reply by bcoconni** [reply]
I am sorry but I am really struggling to see your point. There is no "target" in `c3104.xml` so I don't understand what you mean... The word `target` only appears in the names of the 2 properties `guidance/target_wp_latitude_rad` and `guidance/target_wp_longitude_rad` which define the coordinates of the active waypoint. Please clarify your wording.

In addition, it is unclear if you are talking about the script `c3104.xml`, or a modified version of that script, or the modified X15 with the c310 autopilot that you have already mentioned previously or about your own project (which I gather is some sort of an hypersonic missile/aircraft).

Furthermore, while trying to answer your questions, I brought you **facts** i.e. scripts, XML code, charts showing properties evolution over time,  etc. But when exposing your problem, all we get is a chart which looks like an interpretation of the script results without any **fact** to support your argument. I can't even tell the meaning of the red line and the blue lines. I would like to have property values or anything that I can reproduce myself.
> However, I think these results are not efficient to prove whether the aircraft is heading to the second waypoint or not.

I am afraid I am in strong disagreement with this statement. The chart I showed above is meaningful because the distance decreases **linearly** and that can only happen if the aircraft flies **towards** the waypoint. If the aircraft would have a different heading then the distance would not decrease **linearly** especially as the aircraft is getting closer to the waypoint.
 
Below is a zoomed in version of the chart of the evolution of `guidance/wp-distance` over time while the aircraft is heading towards the second waypoint. As I mentioned earlier the distance (i.e. the radius) decreases linearly which can only happen if the aircraft is heading towards the waypoint.

![wp-distance](https://user-images.githubusercontent.com/7502904/103246352-7a74d880-4963-11eb-883d-664e424b0bb4.jpg)

**Reply by brats2017** [reply]
Thank you so much for your response @bcoconni. My apologies for my unclear explanation. 

> There is no "target" in c3104.xml. Please clarify your wording 

I used "target" for the last waypoint (terminate event: When the aircraft arrives back at Ellington Field). I think the coordinates of this waypoint are:
`  <set name="guidance/target_wp_latitude_rad" value="0.517238"/>`
`  <set name="guidance/target_wp_longitude_rad" value="-1.662727"/>`

> In addition, it is unclear if you are talking about the script c3104.xml, or a modified version of that script, or the modified X15 with the c310 autopilot that you have already mentioned previously or about your own project (which I gather is some sort of an hypersonic missile/aircraft).

In this discussion, I am talking about the script c3104.xml. Not a modified version. Actually, in my own project, I faced a problem, so I need to clarify something in the c3104.xml file to continue working on my project.

> I can't even tell the meaning of the red line and the blue lines. I would like to have property values or anything that I can reproduce myself.

I got this plot using these two properties:
`        <property caption="Latitude       "> position/lat-geod-deg </property>`
`        <property caption="Longitude      "> position/long-gc-deg </property>`
The blue points are the coordinats of the waypoints that is defied in the c3104.xml. The red points are the coordinates of the placemarks.
The lines connect these points respectively. 
Please see the tables that I used to get this plot:
![tables](https://user-images.githubusercontent.com/65332612/103264283-41932e80-4978-11eb-8097-79a6651ffc66.jpg)

When I plot the coordinates of waypoints and placemarks, in the event8, I have a placemark near the terminate (latitude: 0.517238, longitude: -1.662727) while I should have a placemark near the second waypoint (latitude: 0.517533, longitude: -1.663076). 
Also, I think It is possible that I entered the coordinates of the waypoints and placemarks in the table incorrectly, so it causes me to confuse and think there is a problem in the code.

> Below is a zoomed in version of the chart of the evolution of guidance/wp-distance over time while the aircraft is heading towards the second waypoint. 

As I mentioned earlier, I got a similar result for `guidance/wp-distance`. The plot of `guidance/wp-distance` shows the aircraft is heading to the second waypoint because the distance is decreasing. However, when I plot the trajectory of the aircraft, I face a contradiction.

In summary, I want to have a map that shows the waypoints and the trajectory of the c3104.xml. This is my result. My question is why the aircraft is close to the terminate (latitude: 0.517238, longitude: -1.662727) in the second step?
![waypoint1](https://user-images.githubusercontent.com/65332612/103338026-1299bd00-4a4b-11eb-8add-d83c038958b9.jpg)


Can you please plot the trajectory of the c3104.xml and show me your result @bcoconni ? I'm confused because the plot of `guidance/wp-distanc` shows everything is true. However, the trajectory shows contradiction result.
Your help will be greatly appreciated

**Comment by bcoconni** [comment]
I've just migrated this issue to "Discussions" which, I believe, is the right place to manage this topic. This should avoid opening/closing the issue as @brats2017 need support.

**Comment by bcoconni** [comment]
Sorry @brats2017, I just discovered I've missed your last comment. Thanks for clarifying your input. This will need some testing with JSBSim to try to reproduce your test case and understand what's going on. I'll investigate that later this week.

**Reply by brats2017** [reply]
Thank you so much @bcoconni. I will be waiting for your reply. Thanks in advance.

**Reply by brats2017** [reply]
Hello @bcoconni. I wanted to say that my problem has been solved. I read your comments carefully and I realized that I made a mistake when I wanted to plot the trajectory. Consequently, you are right and the code works correctly. My apologies for my mistake.  Thank you so much for your detailed answer.

**Reply by bcoconni** [reply]
> I wanted to say that my problem has been solved.

Excellent ! I'm glad you've been able to find your problem.

**Comment by brats2017** [comment]
Hello all,
I want to set some waypoints with different altitudes for X15. In the c3104.xml the constant altitude has been considered for waypoints. I defined new events for x153.xml and it follows the waypoints in steady altitude; however, I want to design waypoints in different altitudes. Is there any idea on how can I consider altitude as a condition of heading (in addition to latitude and longitude)? 
Thanks in advance

**Comment by seanmcleod** [comment]
But have you looked through `c3104.xml` that you're using? It already has support and examples for altitude hold.

https://github.com/JSBSim-Team/jsbsim/blob/518f2a74d1a5db65de3986816d3e6cd90521585e/scripts/c3104.xml#L25

https://github.com/JSBSim-Team/jsbsim/blob/518f2a74d1a5db65de3986816d3e6cd90521585e/scripts/c3104.xml#L40-L48

**Reply by brats2017** [reply]
Thanks for your explanation. As I understood, c310 when it reachs to 1000 ft and continues its fly in this constant altitude,  follows the waypoints. Consequently, for all of the waypoints, the steady altitude has been considered, but I do not ask x15 to fly in steady altitude. I want to have for example five waypoints that each of them has specific altitude.

**Comment by seanmcleod** [comment]
My point was that the autopilot and the c3104 script have support for setting an altitude target/setpoint, so the basic support is there and there is an example of setting the altitude setpoint to 1000.

So you can modify your version to make use of `ap/altitude_setpoint` and `ap/altitude_hold` to change the altitude setpoint at each waypoint.

**Reply by brats2017** [reply]
Thank you for your explanation. I will try to test that. 
Just one other question is that do you think is it reasonable that without setting  `ap/altitude_setpoin`, ask x15 to head to a waypoint (with specific latitude and longitude)? Or it is necessary to set an altitude for a specific latitude and longitude?
Thanks in advance
